safedelete/admin.py

```diff
@@ -165,3 +167,80 @@ def undelete_selected(self, request, queryset):
                 context,
             )
     undelete_selected.short_description = _("Undelete selected %(verbose_name_plural)s.")
+
+    def hard_delete(self, request, queryset):
+        """ Admin action to hard delete objects. """
+        if not self.has_delete_permission(request):
+            raise PermissionDenied
+
+        original_queryset = queryset.all()
+        undeleted_queryset = queryset.filter(deleted__isnull=True)
+        queryset = queryset.filter(deleted__isnull=False)
+
+        if request.POST.get('post'):
+            requested = original_queryset.count()
+            changed = queryset.count()
+
+            if changed:
+                for obj in queryset:
+                    obj.delete(force_policy=HARD_DELETE)
+                if requested > changed:
+                    self.message_user(
+                        request,
+                        "Successfully hard deleted %(count_changed)d of the "
+                        "%(count_requested)d selected %(items)s." % {
+                            "count_requested": requested,
+                            "count_changed": changed,
+                            "items": model_ngettext(self.opts, requested)
+                        },
+                        messages.WARNING,
+                    )
+                else:
+                    self.message_user(
+                        request,
+                        "Successfully hard deleted %(count)d %(items)s." % {
+                            "count": changed,
+                            "items": model_ngettext(self.opts, requested)
+                        },
+                        messages.SUCCESS,
+                    )
+            else:
+                self.message_user(
+                    request,
+                    "No permission for hard delete. Execute soft delete first.",
```

Comment:
I would change this to something more explicit such as: "You need to soft delete first in order to hard delete an object."